### PR TITLE
FEA Controller for scipy_openblas

### PR DESCRIPTION
**Pull request by jeremiedbb**
Numpy, Scipy, ... will likely be linked against a common build of OpenBLAS for the sci-py ecosystem (https://github.com/MacPython/openblas-libs). At some point it might even become a runtime dependency for these packages.

This special build of OpenBLAS comes with a different naming pattern than OpenBLAS, the lib name being `libscipy_openblas...`, andthe symbols being prefixed by `scipy_`. I made a new controller inherited from the OpenBLAS controller.

I also added a build in the CI to test against the dev versions of numpy and scipy. It would probably have allowed us to face the issue earlier.

### CODE REVIEW

**Comment by jeremiedbb** [comment]
Should I change `internal_api = "openblas"` to `internal_api = "scipy_openblas"` for the new controller ?

**Comment by ogrisel** [comment]
> Should I change internal_api = "openblas" to internal_api = "scipy_openblas" for the new controller ?

I guess so since they do not share common symbols.

**Comment by jeremiedbb** [comment]
The new [pip-dev](https://dev.azure.com/joblib/threadpoolctl/_build/results?buildId=4078&view=logs&jobId=8eeead5b-2ea8-5e0a-9e90-9c963e0feee9&j=8eeead5b-2ea8-5e0a-9e90-9c963e0feee9&t=a860f039-9652-521d-1568-93fd991d4528) CI job confirms that it works:
``tests/test_threadpoolctl.py::test_threadpool_limits_by_prefix[1-libscipy_openblas] PASSED [ 24%]``

**Comment by ogrisel** [comment]
About https://github.com/joblib/threadpoolctl/pull/175#issuecomment-2045659698, I am not so sure. Maybe @rgommers and @mattip have a different opinion. I am not sure about the scope of scipy-openblas: https://github.com/numpy/numpy/issues/26240#issuecomment-2045698774

**Comment by rgommers** [comment]
I commented on the scope in https://github.com/numpy/numpy/issues/26240#issuecomment-2046558865

As for whether this should be a separate controller: it feels a bit wrong conceptually. It is a fairly regular build of OpenBLAS with a symbol prefix set. That's a build-time flag; there are many such build-time flags, and there may be other distributions that do something similar. 

Actually I already see a next change coming: the ILP64 symbol suffix for OpenBLAS is going to become `_64` instead of `64_` at some point. So instead of `openblas_get_config64_` for the regular BLAS controller, it also needs to check for `openblas_get_config_64_` (the trailing `_` may or may not be present).

Probably the more correct way to model this is to consider the symbol naming scheme, which is:
```
<prefix><symbol_name><suffix><compiler-suffix>
```
where:
- `<prefix>` can be nothing or `scipy_`,
- `<symbol_name>` is the plain name as found in the Netlib BLAS/LAPACK docs
- `<suffix>` can be nothing, `_64` or `64_`
- `<compiler-suffix>` can be nothing or `_`

That way this is extensible and wouldn't require separate controllers for different builds of the same library.

**Comment by jeremiedbb** [comment]
I changed to have a same controller for all versions of OpenBLAS. I try do determine first what are the specific prefix, suffix and compiler_suffix and then use them for all symbol lookups. However I face an issue with the above description of the combinatorics of these affixes.

Running `nm -gD /path/to/openblas64 | grep "openblas" shows that ``libopenblas64` already comes with 2 different suffixed versions of some symbols, e.g.
```
0000000000350190 T openblas_set_num_threads_64_
0000000000350fe0 T openblas_set_num_threads64_
```
but the `openblas_set_num_threads_64_` version segfaults.

I can blacklist some combinations but we should take care in future openblas builds to not use these combinations as valid symbol name because it will make things hard for threadpoolctl to support all openblas versions :) For instance, `suffix=_64` + `compiler_suffix=_` is to be avoided.

**Comment by jeremiedbb** [comment]
Actually it's the same with the 32bit int version of openblas:
```
0000000000384ad0 T openblas_set_num_threads
0000000000383c10 T openblas_set_num_threads_
```
and `openblas_set_num_threads_` segfaults. I'll remove the `compiler_suffix` for now.

**Comment by jeremiedbb** [comment]
Ok so looking at OpenBLAS source made it clear, how would have guessed :smile:
``openblas_set_num_threads`` expects an int
``openblas_set_num_threads_`` expects a pointer to an int

File: threadpoolctl.py [hunk 1]
```diff
@@ -149,96 +152,78 @@ def set_num_threads(self, num_threads):
...
-        "openblas_get_config",
-        "openblas_get_config64_",
-        "openblas_get_parallel",
-        "openblas_get_parallel64_",
-        "openblas_get_corename",
-        "openblas_get_corename64_",
+    filename_prefixes = ("libopenblas", "libblas", "libscipy_openblas")
+
+    _sym_prefixes = ("", "scipy_")
+    _sym_suffixes = ("", "64_", "_64")
```
[the remark this answers is not in the document]
Comment (jeremiedbb) [reply threadpoolctl.py]: I'm pretty sure I did that to fit within 88 chars :smile: 

I renamed them

**Pull request merged**